### PR TITLE
implement log opt with containerd logging driver

### PR DESCRIPTION
**Pull request by fahedouch**
- using containerd runtime v2 logging plugin mode.
- add log-opt for run subcmd 

Signed-off-by: fahed dorgaa <fahed.dorgaa@gmail.com>

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
What's this?

**Comment by fahedouch** [comment]
@AkihiroSuda  it is a draft PR to fix  https://github.com/containerd/nerdctl/issues/864, I started working on this. My code is on local for the moment!

File: pkg/taskutil/taskutil.go [hunk 1]
```diff
@@ -45,14 +47,23 @@ func NewTask(ctx context.Context, client *containerd.Client, container container
...
 			return nil, err
 		}
-		ioCreator = cio.LogURI(u)
+		//using containerd runtime v2 logging plugin mode to handle task log entries
+		q := u.Query()
+		fn, err := loggingpkg.GetLoggerFunc(q.Get(loggingpkg.MagicArgv1))
+		if err != nil {
+			return nil, err
+		}
+		logging.Run(fn)
```
Comment (AkihiroSuda) [inline pkg/taskutil/taskutil.go:56]: This function can only be called from the logging plugin

Comment (fahedouch) [reply pkg/taskutil/taskutil.go]: got it

**Comment by fahedouch** [comment]
windows job fails.. :/

File: README.md [hunk 2]
```diff
@@ -472,6 +472,9 @@ Metadata flags:
 - :whale: :blue_square: `--cidfile`: Write the container ID to the file
 - :nerd_face: `--pidfile`: file path to write the task's pid. The CLI syntax conforms to Podman convention.
 
+Logging flags:
+- :whale: `--log-opt`: Set containerd runtime v2 logging plugin options
```
Comment (AkihiroSuda) [inline README.md:476]: Please explain the implemented flags

File: cmd/nerdctl/run_test.go [hunk 3]
```diff
@@ -198,3 +199,43 @@ func TestRunEnv(t *testing.T) {
 		return nil
 	})
 }
+
+func TestRunWithLogOpt(t *testing.T) {
+	t.Parallel()
+	// docker use a different driver to support logging option
+	testutil.DockerIncompatible(t)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_test.go:206]: Why incompatible?
https://docs.docker.com/config/containers/logging/json-file/

File: cmd/nerdctl/run_test.go [hunk 4]
```diff
@@ -198,3 +199,43 @@ func TestRunEnv(t *testing.T) {
...
+	}
+	base := testutil.NewBase(t)
+	containerName := testutil.Identifier(t)
+
+	defer base.Cmd("rm", "-f", containerName).AssertOK()
+	base.Cmd("run", "-d", "--log-opt", "max-size=5k", "--log-opt", "max-files=2", "--name", containerName, testutil.CommonImage,
+		"sh", "-euxc", "du").AssertOK()
+
+	time.Sleep(3 * time.Second)
+	// Get log JSONFile path
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_test.go:218]: `base.InspectContainer(containerName).LogPath` contains the json file path.

File: cmd/nerdctl/run_test.go [hunk 5]
```diff
@@ -198,3 +199,43 @@ func TestRunEnv(t *testing.T) {
...
+	testutil.DockerIncompatible(t)
+	if runtime.GOOS == "windows" {
+		t.Skip("`logging options` are not yet implemented on Windows")
+	}
+	base := testutil.NewBase(t)
+	containerName := testutil.Identifier(t)
+
+	defer base.Cmd("rm", "-f", containerName).AssertOK()
+	base.Cmd("run", "-d", "--log-opt", "max-size=5k", "--log-opt", "max-files=2", "--name", containerName, testutil.CommonImage,
+		"sh", "-euxc", "du").AssertOK()
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_test.go:215]: Output of "du" doesn't seem enough to test the log rotation.

Comment (AkihiroSuda) [reply cmd/nerdctl/run_test.go:215]: `hexdump -C /dev/urandom  | head -nXXXX` is an easy way to generate random long text output

File: cmd/nerdctl/run_test.go (hunk 3 above)
Comment (fahedouch) [reply cmd/nerdctl/run_test.go]: fixed

File: README.md [hunk 6]
```diff
@@ -472,6 +472,13 @@ Metadata flags:
 - :whale: :blue_square: `--cidfile`: Write the container ID to the file
 - :nerd_face: `--pidfile`: file path to write the task's pid. The CLI syntax conforms to Podman convention.
 
+Logging flags:
+- :whale: `--log-driver=(local|json-file|fluentd|fluentbit)`: Logging driver for the container (default `json-file`).
```
Comment (AkihiroSuda) [inline README.md:476]: Unimplemented ones shouldn't be documented here.

File: cmd/nerdctl/run_test.go [hunk 7]
```diff
@@ -212,3 +213,35 @@ func TestRunStdin(t *testing.T) {
...
+	base.Cmd("run", "-d", "--log-driver", "json-file", "--log-opt", "max-size=5K", "--log-opt", "max-file=2", "--name", containerName, testutil.CommonImage,
+		"sh", "-euxc", "hexdump -C /dev/urandom | head -n1000").AssertOK()
+
+	time.Sleep(3 * time.Second)
+	inspectedContainer := base.InspectContainer(containerName)
+	logJSONPath := filepath.Dir(inspectedContainer.LogPath)
+	// matches = current log file + old log files to retain
+	matches, err := filepath.Glob(filepath.Join(logJSONPath, inspectedContainer.ID+"*"))
+	assert.NilError(t, err)
+	if len(matches) > 3 {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_test.go:234]: len(matches) == 2

**Review by junnplus — APPROVED** [review]
LGTM

**Review by AkihiroSuda — APPROVED** [review]
Thanks

File: cmd/nerdctl/main.go [hunk 8]
```diff
@@ -71,10 +71,22 @@ func main() {
...
 
 func xmain() error {
-	if len(os.Args) == 3 && os.Args[1] == logging.MagicArgv1 {
+	loggingMode := false
+	for _, arg := range os.Args {
+		if arg == logging.MagicArgv1 {
+			loggingMode = true
+			break
+		}
+	}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/main.go:80]: Sorry for going back and forth, but this logic doesn’t seem robust.
We have to ensure that MagicArgv1 is always argv1.
Other args should be encoded in argv2 or in the metadata referred by argv2.

**Pull request merged**